### PR TITLE
Migrate to flight auth

### LINKED ISSUE #3513: Python client should migrate to flight auth
The SessionService.NewSession grpc call is deprecated. Instead either use the FlightService.Handshake method (aka Flight Auth v1) or set the authorization header on the first request to get a bearer token to use on later requests.

Likewise, instead of SessionService.RefreshSessionToken, when the server's response includes another authorization header, use that for future requests.

See https://github.com/deephaven/deephaven-core/tree/main/authentication for more details.

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #3513

### CODE REVIEW

File: py/client/pydeephaven/session.py [hunk 1]
```diff
@@ -60,7 +115,15 @@ def __init__(self, host: str = None, port: int = None, never_timeout: bool = Tru
             self.port = int(os.environ.get("DH_PORT", 10000))
 
         self.is_connected = False
-        self.session_token = None
+
+        if auth_type == "Anonymous":
+            self._auth_token = auth_type
+        elif auth_type == "Basic":
+            auth_token_base64 = base64.b64encode(auth_token.encode("ascii")).decode("ascii")
```
Comment (chipkent) [inline py/client/pydeephaven/session.py:122]: @niloc132 If the Go client should be split up into two args, it probably also needs to base64 encode...right?

File: py/client/pydeephaven/session.py [hunk 2]
```diff
@@ -86,11 +152,12 @@ def __exit__(self, exc_type, exc_val, exc_tb):
     def tables(self):
         with self._r_lock:
             fields = self._fetch_fields()
-            return [field.field_name for field in fields if field.application_id == 'scope' and field.typed_ticket.type == 'Table']
+            return [field.field_name for field in fields if
+                    field.application_id == 'scope' and field.typed_ticket.type == 'Table']
 
     @property
     def grpc_metadata(self):
```
Comment (chipkent) [inline py/client/pydeephaven/session.py:159]: None of these methods have type hints.  The client needs a major lookover.

File: py/client/pydeephaven/session.py [hunk 3]
```diff
@@ -124,6 +191,13 @@ def app_service(self):
...
         return self._app_service
 
+    @property
+    def config_service(self):
+        if not self._config_service:
+            self._config_service = ConfigService(self)
+
+        return self._config_service
+
     def make_ticket(self, ticket_no=None):
```
Comment (chipkent) [inline py/client/pydeephaven/session.py:201]: private?
The client API needs to look at all of the methods and confirm if they should be public or not.

I _think_ that the tickets are not exposed to users in the Go client.  Should they be in python?  ... again, something for another PR.

File: py/client/pydeephaven/session.py [hunk 4]
```diff
@@ -153,12 +227,27 @@ def _fetch_fields(self):
 
     def _connect(self):
         with self._r_lock:
-            self.grpc_channel, self.session_token, self._timeout = self.session_service.connect()
-            self.is_connected = True
+            try:
+                self._flight_client = paflight.connect(location=(self.host, self.port), middleware=[
+                    _DhClientAuthMiddlewareFactory(self)])
+                self._auth_handler = _DhClientAuthHandler(self)
+                self._flight_client.authenticate(self._auth_handler)
```
Comment (chipkent) [inline py/client/pydeephaven/session.py:234]: Is the initial auth only used once?  I think it is only being used once.  @niloc132 mentioned problems if it gets reused.

File: py/client/pydeephaven/session.py [hunk 5]
```diff
@@ -169,9 +258,10 @@ def _keep_alive(self):
     def _refresh_token(self):
```
Comment (chipkent) [inline py/client/pydeephaven/session.py:258]: See https://github.com/deephaven/deephaven-core/issues/3593.  @jcferretti can comment further.

File: py/client/pydeephaven/session.py [hunk 6]
```diff
@@ -201,6 +291,7 @@ def close(self) -> None:
                 self.grpc_channel.close()
                 self.is_connected = False
                 self._last_ticket = 0
+                self._flight_client.close()
                 # self._executor.shutdown()
```
Comment (chipkent) [inline py/client/pydeephaven/session.py:295]: needed?

File: py/client/pydeephaven/session.py [hunk 7]
```diff
@@ -124,6 +191,13 @@ def app_service(self):
 
         return self._app_service
 
+    @property
+    def config_service(self):
```
[the remark this answers is not in the document]
Comment (chipkent) [reply py/client/pydeephaven/session.py:195]: I see the same problem with other services.  It needs significant cleanup... probably in another PR.

File: py/client/pydeephaven/session.py [hunk 8]
```diff
@@ -153,12 +227,27 @@ def _fetch_fields(self):
 
     def _connect(self):
         with self._r_lock:
```
Comment (chipkent) [inline py/client/pydeephaven/session.py:229]: Where the lock is created, I would comment about what the lock is for.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/client/pydeephaven/_config_service.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+from pydeephaven.dherror import DHError
+from pydeephaven.proto import config_pb2, config_pb2_grpc
+
+
+class ConfigService:
+    def __init__(self, session):
+        self.session = session
+        self._grpc_app_stub = config_pb2_grpc.ConfigServiceStub(session.grpc_channel)
+
+    def get_configuration_constants(self) -> dict:
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply py/client/pydeephaven/_config_service.py:14]: Yes, the key should be str, but the value, though it is str right now,  could be any type as the comment for the value field in config.proto explicitly made clear.

File: py/client/pydeephaven/session.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jmao-denver) [reply py/client/pydeephaven/session.py:195]: The original reason why these are not private and don't have docstrings is that 
1. they don't show up in the sphinx generated docs. 
2. the static checker in PyCharm would complain about accessing a private property in other modules

File: py/client/pydeephaven/session.py (hunk 3 above)
Comment (jmao-denver) [reply py/client/pydeephaven/session.py:201]: see the reason given above.

File: py/client/pydeephaven/session.py (hunk 2 above)
Comment (jmao-denver) [reply py/client/pydeephaven/session.py:159]: see response above.

File: py/client/pydeephaven/session.py (hunk 4 above)
Comment (jmao-denver) [reply py/client/pydeephaven/session.py:234]: yes.

File: py/client/pydeephaven/session.py [hunk 10]
```diff
@@ -35,12 +83,19 @@ class Session:
...
 
         Args:
             host (str): the host name or IP address of the remote machine, default is 'localhost'
             port (int): the port number that Deephaven server is listening on, default is 10000
+            auth_type (str): the authentication type string, can be "Anonymous', 'Basic", or any custom-built
+                authenticator in the server, such as "io.deephaven.authentication.psk.PskAuthenticationHandler",
+                default is 'Anonymous'.
+            auth_token (str): the authentication token string. When auth_type is 'Basic', it must be
+                "user:password"; when auth_type is "Anonymous', it will be ignored; when auth_type is a custom-built
+                authenticator, it must conform to the specific requirement of the authenticator
```
[the remark this answers is not in the document]
Comment (niloc132) [reply py/client/pydeephaven/session.py:98]: They syntax of the Authorization header is in two parts, the "scheme" and "parameters", see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization#syntax

See also
https://httpwg.org/specs/rfc9110.html#field.authorization
and
https://httpwg.org/specs/rfc9110.html#credentials

Sending it as one value is probably fine, but most use cases will tend to think of it as two, likely a constant for the kind of authentication they are using, then the specific token that identifies them.

File: py/client/pydeephaven/session.py (hunk 1 above)
Comment (niloc132) [reply py/client/pydeephaven/session.py:122]: Only the `Basic` type is expected/required to be base64 encoded, this isn't necessarily required for other formats. For example, if you base64 the already-hex encoded uuids that are used for bearer tokens, or the already-base64 encoded openId connect payloads, the server won't be able to read them.

**Review by chipkent — APPROVED** [review]

**Review by chipkent — DISMISSED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

How-to: https://github.com/deephaven/deephaven.io/issues/2369
Conceptual: https://github.com/deephaven/deephaven.io/issues/2366
Reference: https://github.com/deephaven/deephaven.io/issues/2368
Blog: https://github.com/deephaven/deephaven.io/issues/2367

**Pull request merged**